text/0000-proc-macro-include.md

```diff
@@ -0,0 +1,205 @@
+
+# Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- [`proc_macro::tracked_path`](https://doc.rust-lang.org/stable/proc_macro/tracked_path/fn.path.html) (unstable)
+
+This just tells the proc_macro driver that the proc macro has a dependency on the given path.
+This is sufficient for tracking the file, as the proc macro can just also read the file itself,
+but lacks the ability to require the proc macro go through this API, or to provide spans for errors.
+
+Meaningfully, it'd be nice to be able to sandbox proc macros in wasm à la [watt](https://crates.io/crates/watt)
+while still having proc macros capable of reading the filesystem (in a proc_macro driver controlled manner).
+
+- Custom error type
+
+A custom error wrapper would provide a point to attach more specific error information than just an
+`io::Error`, such as the lexer error encountered by `include`. This RFC opts to use `io::Error`
+directly to provide a more minimal API surface.
+
+- Wrapped return types
+
+Returning `Literal::string` from `include_str` and `Vec<u8>` from `include_bytes` implies that
+the entire included file must be read into memory managed by the Rust global allocator.
+Alternatively, a more abstract buffer type could be used which allows more efficiently working
+with very large files that could be instead e.g. memmapped rather than read into a buffer.
+
+This would likely look like `LiteralString` and `LiteralBytes` types in the `proc_macro` bridge,
+but this RFC opts to use the existing `Literal` and `Vec<u8>` to provide a more minimal API surface.
+
+- Status quo
+
+Proc macros can continue to read files and use `include_str!` to indicate a build dependency.
+This is error prone, easy to forget to do, and all around not a great experience.
+
+# Prior art
+[prior-art]: #prior-art
+
+No known prior art.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- It would be nice for `include` to allow emitting a useful lexer error directly.
+  This is not currently provided for by the proposed API.
+- `include!` sets the "current module path" for the included code.
+  It's unclear how this should behave for `proc_macro::include`,
+  and whether this behavior should be replicated at all.
+- Should `include_str` get source code normalization (i.e. `\r\n` to `\n`)?
+  `include_str!` deliberately includes the string exactly as it appears on disk,
+  and the purpose of these APIs is to provide post-processing steps,
+  which could need the file to be reproduced exactly,
+  so the answer is likely *no*,
+  and the produced `Literal` should represent the exact contents of the file.
+- What base directory should relative paths be resolved from?
+  The two reasonable answers are
+
+  - That which `include!` is relative to in the source file expanding the macro.
+  - That which `fs` is relative to in the proc macro execution.
+
+  Both have their merits and drawbacks.
```

Comment:
One way to support both options would be to take a `Span` that the path is relative to. Then it would make multi-level includes easier (the macro includes a path relative to the Rust source file, then the included file references _another_ relative file so that needs to be included based on the `Span` from the first `proc_macro::include_str` call).

Reply:
I suppose it should just behave the exact same as a `include_str!("..")` macro invocation whose tokens carry a mixed_site span.